src/smt/theory_str_noodler/theory_str_noodler.cpp

```diff
@@ -1889,6 +1893,101 @@ namespace smt::noodler {
+        // x <= y -> x = y | x < y
+        add_axiom({~lit_e, lit_e_lt, lit_x_y});
+    }
+
+    /**
+     * @brief Handle str.<
+     * Translates to the following theory axioms.
+     * 
+     * x = eps & y != eps -> x < y
+     * x < y & x != eps -> x = u.v1.w1
+     * x < y & x != eps -> v1 in re.allchar
+     * x < y & x != eps -> v2 in re.allchar
+     * x < y & x != eps -> to_code(v1) + k = to_code(v2) & k >= 1
+     * @param e str.< predicate
+     */
+    void theory_str_noodler::handle_lex_lt(expr *e) {
+        if(axiomatized_persist_terms.contains(e))
+            return;
+
+        axiomatized_persist_terms.insert(e);
+        STRACE("str", tout  << "handle str.< " << mk_pp(e, m) << std::endl;);
+
+        expr *x = nullptr, *y = nullptr;
+        VERIFY(m_util_s.str.is_lt(e, x, y));
+        expr_ref eps(m_util_s.str.mk_string(""), m);
+        expr_ref x_eps(m.mk_eq(x, eps), m);
+        expr_ref y_eps(m.mk_eq(y, eps), m);
+
+        expr_ref lex_pre = mk_str_var_fresh("lex_pre");
+        expr_ref lex_in_left = mk_str_var_fresh("lex_in_left");
+        expr_ref lex_in_right = mk_str_var_fresh("lex_in_right");
+        expr_ref lex_post_left = mk_str_var_fresh("lex_post_left");
+        expr_ref lex_post_right = mk_str_var_fresh("lex_post_right");
+        expr_ref px(m_util_s.str.mk_concat(m_util_s.str.mk_concat(lex_pre, lex_in_left), lex_post_left), m);
+        expr_ref py(m_util_s.str.mk_concat(m_util_s.str.mk_concat(lex_pre, lex_in_right), lex_post_right), m);
+        string_theory_propagation(px);
+        string_theory_propagation(py);
+
+        expr_ref x_px(m.mk_eq(x, px), m);
+        expr_ref y_py(m.mk_eq(y, py), m);
+        literal lit_e = mk_literal(e);
+        literal lit_x_px = mk_literal(x_px);
+        literal lit_y_py = mk_literal(y_py);
+        
+        expr_ref re_in_left(m_util_s.re.mk_in_re(lex_in_left, m_util_s.re.mk_full_char(nullptr)), m);
+        expr_ref re_in_right(m_util_s.re.mk_in_re(lex_in_right, m_util_s.re.mk_full_char(nullptr)), m);
+        expr_ref to_code_left(m_util_s.str.mk_to_code(lex_in_left), m);
+        expr_ref to_code_right(m_util_s.str.mk_to_code(lex_in_right), m);
+  
+        // This is a dirty hack. If I add axiom to_code(v1) < to_code(v2), the LIA solver starts 
+        // to solve a nonlinear problem (?). If I use to_code(v1) + k = to_code(v2) where k > 0, it works well.
+        expr_ref vark = mk_int_var_fresh("lex_add");
+        expr_ref to_code_lt(m.mk_eq(m_util_a.mk_add(to_code_left, vark), to_code_right), m);
+        // k >= 0
+        add_axiom({mk_literal(m_util_a.mk_ge(vark, m_util_a.mk_int(1)))});
+        
+        literal lit_x_eps = mk_literal(x_eps);
+        literal lit_y_eps = mk_literal(y_eps);
+
+        // x = eps & y != eps -> x < y
+        add_axiom({~lit_x_eps, lit_y_eps, lit_e});
+        // x < y & x != eps -> x = u.v1.w1
+        add_axiom({~lit_e, lit_x_eps, lit_x_px});
+        // x < y & x != eps -> y = u.v2.w2
```

Comment:
This line is missing in the function documentation.

Reply:
Fixed